core/src/main/java/com/linecorp/armeria/common/stream/TimeoutStreamMessage.java

```diff
@@ -0,0 +1,270 @@
+            this.executor = requireNonNull(executor, "executor");
+            this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+            timeoutNanos = timeoutDuration.toNanos();
+            this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+        }
+
+        private ScheduledFuture<?> scheduleTimeout(long delay, TimeUnit unit) {
+            return executor.schedule(this, delay, unit);
+        }
+
+        void cancelSchedule() {
+            if (timeoutFuture != null && !timeoutFuture.isCancelled()) {
+                timeoutFuture.cancel(false);
+            }
+        }
+
+        @Override
+        public void run() {
+            if (timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+                final long currentTimeNanos = System.nanoTime();
+                final long elapsedNanos = currentTimeNanos - lastEventTimeNanos;
+
+                if (elapsedNanos < timeoutNanos) {
+                    final long delayNanos = timeoutNanos - elapsedNanos;
+                    timeoutFuture = scheduleTimeout(delayNanos, TimeUnit.NANOSECONDS);
+                    return;
+                }
+            }
+            completed = true;
+            delegate.onError(new TimeoutException(
+                    String.format(TIMEOUT_MESSAGE, timeoutDuration.toMillis(), timeoutMode)));
+            subscription.cancel();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            subscription = s;
+            delegate.onSubscribe(this);
+            if (completed) {
+                return;
+            }
+            lastEventTimeNanos = System.nanoTime();
+            timeoutFuture = scheduleTimeout(timeoutNanos, TimeUnit.NANOSECONDS);
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (completed) {
+                return;
+            }
+            switch (timeoutMode) {
+                case UNTIL_NEXT:
+                    lastEventTimeNanos = System.nanoTime();
+                    break;
+                case UNTIL_FIRST:
+                    cancelSchedule();
+                    break;
+                case UNTIL_EOS:
+                    break;
+            }
+            delegate.onNext(t);
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            if (completed) {
+                return;
+            }
+            completed = true;
+            cancelSchedule();
+            delegate.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
+            if (completed) {
+                return;
+            }
+            completed = true;
+            cancelSchedule();
+            delegate.onComplete();
+        }
+
+        @Override
+        public void request(long l) {
+            subscription.request(l);
+        }
+
+        @Override
+        public void cancel() {
+            completed = true;
```

Comment:
If the downstream subscriber decided to call `TimeoutSubscriber.cancel` directly, it should still receive a `onError` or `onComplete` callback

```suggestion
```